### PR TITLE
Add arbitrary labels to docker images

### PR DESCRIPTION
**Pull request by mmilata**
Trello card: https://trello.com/c/HB9IwGj5/925-5-add-arbitrary-labels-to-docker-images
Documentation: openshift/openshift-docs#3015

Missing:
- [x] needs openshift/source-to-image#601 merged
- [x] admission plugin implementation
- [x] integration/extended tests

### CODE REVIEW

File: pkg/build/api/v1/types.go [hunk 1]
```diff
@@ -622,6 +622,19 @@ type BuildOutput struct {
 	// up the authentication for executing the Docker push to authentication
 	// enabled Docker Registry (or Docker Hub).
 	PushSecret *kapi.LocalObjectReference `json:"pushSecret,omitempty" protobuf:"bytes,2,opt,name=pushSecret"`
+
+	// labels define a list of labels that are applied to the resulting image. If there are
+	// multiple labels with the same name then the last one in the list is used.
```
Comment (liggitt) [inline pkg/build/api/v1/types.go]: validation should complain about multiple labels with the same name

File: pkg/build/api/v1/types.go [hunk 2]
```diff
@@ -622,6 +622,19 @@ type BuildOutput struct {
...
+	Labels []ImageLabel `json:"labels,omitempty" protobuf:"bytes,3,rep,name=labels"`
```
Comment (liggitt) [inline pkg/build/api/v1/types.go]: it's confusing that these will result in docker labels, not API labels on the object that the `to` reference points to

File: pkg/build/api/v1/types.go (hunk 1 above)
Comment (mmilata) [inline pkg/build/api/v1/types.go]: The idea was that the BuildDefaults admission plugin will prepend labels to the list making them effectively defaults and BuildOverrides will append them making them mandatory. Not really a problem though, this logic can be extended to remove the duplicate entry.

File: pkg/build/api/v1/types.go (hunk 2 above)
Comment (mmilata) [inline pkg/build/api/v1/types.go]: Should I rename the field to `imageLabels` or `dockerLabels`?

Comment (liggitt) [inline pkg/build/api/v1/types.go]: I'd vote for `imageLabels`, curious what others think

**Comment by mmilata** [comment]
updated:
* implemented admission plugins
* address issues found in review

File: pkg/build/api/v1/types.go (hunk 2 above)
Comment (mmilata) [inline pkg/build/api/v1/types.go]: Renamed the field to `imageLabels`.

**Comment by mmilata** [comment]
Added simple extended test.

**Comment by mmilata** [comment]
Documentation: https://github.com/openshift/openshift-docs/pull/3015

**Comment by mmilata** [comment]
@bparees PTAL

File: pkg/build/admission/defaults/admission_test.go [hunk 3]
```diff
@@ -130,3 +132,137 @@ func TestIncrementalDefaults(t *testing.T) {
...
+				{
+					Name:  "distribution-scope",
+					Value: "private",
+				},
+				{
+					Name:  "changelog-url",
+					Value: "file:///dev/null",
+				},
+			},
+		},
```
Comment (bparees) [inline pkg/build/admission/defaults/admission_test.go]: needs a test where the build object has no labels so you can confirm you can successfully add labels to a build definition that has an empty label field.

Comment (bparees) [inline pkg/build/admission/defaults/admission_test.go]: nm, i see that's the first test.

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
[testextended][extended:core(builds)]

**Comment by bparees** [comment]
@mmilata please squash the 2 non-bump commits
@openshift/api-review please review/sign off.

**Comment by smarterclayton** [comment]
API approved

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/11240
[merge]

**Comment by mmilata** [comment]
@bparees, squashed & rebased on current master

**Comment by bparees** [comment]
[merge]

**Comment by mmilata** [comment]
Both failures seem to be caused by AWS capacity problem:

> InsufficientInstanceCapacity => We currently do not have sufficient m4.xlarge capacity in the Availability Zone you requested (us-east-1d). Our system will be working on provisioning additional capacity. You can currently get m4.xlarge capacity by not specifying an Availability Zone in your request or choosing us-east-1c, us-east-1b.

**Comment by mmilata** [comment]
looks like flake  #11024

**Comment by mfojtik** [comment]
[merge]

**Comment by mmilata** [comment]
@mfojtik, why is the bot ignoring this?:)

**Comment by mfojtik** [comment]
@mmilata it is not, see: https://github.com/openshift/origin/pull/11209#issuecomment-252546013

**Pull request merged**